### PR TITLE
Fixed #24295 -- Allowed ModelForm meta to specify field classes.

### PR DESCRIPTION
**Pull request by loic**

### CODE REVIEW

File: django/forms/models.py [hunk 1]
```diff
@@ -205,6 +208,8 @@ def fields_for_model(model, fields=None, exclude=None, widgets=None,
             kwargs['help_text'] = help_texts[f.name]
         if error_messages and f.name in error_messages:
             kwargs['error_messages'] = error_messages[f.name]
+        if field_classes and f.name in field_classes:
+            kwargs['form_class'] = field_classes[f.name]
```
Comment (MarkusH) [inline django/forms/models.py]: Shouldn't that be `'field_class'` as key?

Comment (carljm) [inline django/forms/models.py]: No, the `form_class` kwarg to `formfield()` actually means a form field class.

File: django/forms/models.py [hunk 2]
```diff
@@ -176,8 +180,7 @@ def fields_for_model(model, fields=None, exclude=None, widgets=None,
     ``error_messages`` is a dictionary of model field names mapped to a
     dictionary of error messages.
 
-    ``formfield_callback`` is a callable that takes a model field and returns
-    a form field.
+    ``field_classes`` is a dictionary of model field names mapped to a field class.
```
Comment (carljm) [inline django/forms/models.py]: "mapped to a form field class"

File: docs/topics/forms/modelforms.txt [hunk 3]
```diff
@@ -525,9 +524,10 @@ the ``name`` field::
                 },
             }
 
-Finally, if you want complete control over of a field -- including its type,
-validators, etc. -- you can do this by declaratively specifying fields like you
-would in a regular ``Form``.
+.. versionadded:: 1.8
+
+You can also specify ``field_classes`` to customize the type of fields
+instanciated by the form.
```
Comment (carljm) [inline docs/topics/forms/modelforms.txt]: instantiated

Comment (carljm) [inline docs/topics/forms/modelforms.txt]: I think usually we try to describe the feature integrated into the main text, and then just add a versionadded note like "The `Meta.field_classes` argument was added" (and I think that text should actually be indented under the versionadded directive). This makes removing versionadded directives simpler later on.

File: django/forms/models.py (hunk 1 above)
Comment (MarkusH) [inline django/forms/models.py]: Got it. Thanks

**Comment by loic** [comment]
@carljm @timgraham can I ship it?

File: docs/ref/forms/models.txt [hunk 4]
```diff
@@ -64,7 +67,11 @@ Model Form API reference. For introductory material about model forms, see the
 
     See :ref:`model-formsets` for example usage.
 
-.. function:: inlineformset_factory(parent_model, model, form=ModelForm, formset=BaseInlineFormSet, fk_name=None, fields=None, exclude=None, extra=3, can_order=False, can_delete=True, max_num=None, formfield_callback=None, widgets=None, validate_max=False, localized_fields=None, labels=None, help_texts=None, error_messages=None, min_num=None, validate_min=False)
+    .. versionadded:: 1.8
```
Comment (timgraham) [inline docs/ref/forms/models.txt]: 1.9 (and missing this note for `modelform_factory`)

File: docs/topics/forms/modelforms.txt [hunk 5]
```diff
@@ -475,9 +475,8 @@ Overriding the default fields
 
 The default field types, as described in the `Field types`_ table above, are
 sensible defaults. If you have a ``DateField`` in your model, chances are you'd
-want that to be represented as a ``DateField`` in your form. But
-``ModelForm`` gives you the flexibility of changing the form field type and
-widget for a given model field.
```
Comment (timgraham) [inline docs/topics/forms/modelforms.txt]: The reasoning for removing widget here isn't immediately obvious to me.

Comment (loic) [inline docs/topics/forms/modelforms.txt]: I didn't think it was that helpful to single out widgets here, "changing the form field for a given model" includes customising the widget. The example for overriding widget starts two lines below, then followed by all the other attributes that can be overridden in the same way.

You want it back?

File: docs/ref/forms/models.txt (hunk 4 above)
Comment (loic) [inline docs/ref/forms/models.txt]: Fixed!

File: docs/topics/forms/modelforms.txt (hunk 5 above)
Comment (timgraham) [inline docs/topics/forms/modelforms.txt]: Sounds fine.

**Pull request merged**